File: pubnub-java/src/integrationTest/java/com/pubnub/api/integration/GroupManagementIntegrationTests.java
package com.pubnub.api.integration;

import com.pubnub.api.integration.util.BaseIntegrationTest;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.CountDownLatch;

import static com.pubnub.api.integration.util.Utils.random;
import static com.pubnub.api.integration.util.Utils.randomChannel;
import static org.junit.Assert.*;

public class GroupManagementIntegrationTests extends BaseIntegrationTest {

    private String mChannel1;
    private String mChannel2;
    private String mChannel3;

    private String mGroup;

    @Override
    protected void onBefore() {
        mChannel1 = randomChannel();
        mChannel2 = randomChannel();
        mChannel3 = randomChannel();
        mGroup = "cg_".concat(random());
    }

    @Override
    protected void onAfter() {

    }

    @Test
    public void testRemoveChannelsFromGroup() throws InterruptedException {
        final CountDownLatch signal = new CountDownLatch(1);

        addChannelsToGroup();

        pubNub.removeChannelsFromChannelGroup()
                .channelGroup(mGroup)
                .channels(Arrays.asList(mChannel1, mChannel2, mChannel3))
                .async((result, status) -> {
                    assertFalse(status.isError());
                    assert status.getAffectedChannels() != null;
                    assertEquals(3, status.getAffectedChannels().size());
                    assertEquals(0, pubNub.getSubscribedChannels().size());
                    assert status.getAffectedChannelGroups() != null;
                    assertEquals(1, status.getAffectedChannelGroups().size());
                    assertEquals(0, pubNub.getSubscribedChannelGroups().size());
                    signal.countDown();
                });

        signal.await();
    }

    @Test
    public void testRemoveChannelFromGroup() throws InterruptedException {
        final CountDownLatch signal = new CountDownLatch(1);

        addChannelsToGroup();

        pubNub.removeChannelsFromChannelGroup()
                .channelGroup(mGroup)
                .channels(Collections.singletonList(mChannel1))
                .async((result, status) -> {
                    assertFalse(status.isError());
                    signal.countDown();
                });

        signal.await();
    }

    @Test
    public void testSubscribeToChannelGroup() throws InterruptedException {
        addChannelsToGroup();
        subscribeToChannelGroup(pubNub, mGroup);

        boolean isGroupSubscribed = false;
        for (int i = 0; i < pubNub.getSubscribedChannelGroups().size(); i++) {
            if (pubNub.getSubscribedChannelGroups().get(i).equals(mGroup)) {
                isGroupSubscribed = true;
            }
        }

        assertTrue(isGroupSubscribed);
    }

    @Test
    public void testUnsubscribeFromChannelGroup() throws InterruptedException {
        addChannelsToGroup();
        subscribeToChannelGroup(pubNub, mGroup);

        pubNub.unsubscribe()
                .channelGroups(Collections.singletonList(mGroup))
                .execute();

        pause(1);

        assertEquals(0, pubNub.getSubscribedChannelGroups().size());
    }

    @Test
    public void testGetAllChannelsFromGroup() throws InterruptedException {
        final CountDownLatch signal = new CountDownLatch(1);

        addChannelsToGroup();

        pubNub.listChannelsForChannelGroup()
                .channelGroup(mGroup)
                .async((result, status) -> {
                    assertFalse(status.isError());
                    assert result != null;
                    assertEquals(3, result.getChannels().size());
                    signal.countDown();
                });

        signal.await();
    }

    @Test
    public void testAddChannelsToGroup() throws InterruptedException {
        final CountDownLatch signal = new CountDownLatch(1);

        pubNub.addChannelsToChannelGroup()
                .channelGroup(mGroup)
                .channels(Arrays.asList(mChannel1, mChannel2, mChannel3))
                .async((result, status) -> {
                    assertFalse(status.isError());
                    assert status.getAffectedChannelGroups() != null;
                    assertEquals(1, status.getAffectedChannelGroups().size());
                    assert status.getAffectedChannels() != null;
                    assertEquals(3, status.getAffectedChannels().size());
                    signal.countDown();
                });

        signal.await();
    }

    private void addChannelsToGroup() throws InterruptedException {
        final CountDownLatch signal = new CountDownLatch(1);

        pubNub.addChannelsToChannelGroup()
                .channelGroup(mGroup)
                .channels(Arrays.asList(mChannel1, mChannel2, mChannel3))
                .async((result, status) -> {
                    assertFalse(status.isError());
                    signal.countDown();
                });

        signal.await();
    }

}

File: core/src/integrationTest/kotlin/com/pubnub/api/integration/GroupManagementIntegrationTests.kt
package com.pubnub.api.integration

import com.pubnub.api.CommonUtils.randomValue
import com.pubnub.api.CommonUtils.retry
import com.pubnub.api.await
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class GroupManagementIntegrationTests : BaseIntegrationTest() {

    lateinit var expectedChannel1: String
    lateinit var expectedChannel2: String
    lateinit var expectedChannel3: String

    lateinit var expectedGroup: String

    override fun onBefore() {
        expectedChannel1 = randomValue()
        expectedChannel2 = randomValue()
        expectedChannel3 = randomValue()
        expectedGroup = randomValue(8)
    }

    @Test
    fun testRemoveChannelsFromGroup() {
        addChannelsToGroup()

        pubnub.removeChannelsFromChannelGroup(
            channelGroup = expectedGroup,
            channels = listOf(expectedChannel1, expectedChannel2, expectedChannel3)
        ).await { _, status ->
            assertFalse(status.error)
            assertEquals(3, status.affectedChannels.size)
            assertEquals(0, pubnub.getSubscribedChannels().size)
            assertEquals(1, status.affectedChannelGroups.size)
            assertEquals(0, pubnub.getSubscribedChannelGroups().size)
        }
    }

    @Test
    fun testRemoveChannelFromGroup() {
        addChannelsToGroup()

        pubnub.removeChannelsFromChannelGroup(
            channelGroup = expectedGroup,
            channels = listOf(expectedChannel1)
        ).await { _, status ->
            assertFalse(status.error)
        }
    }

    @Test
    fun testSubscribeToChannelGroup() {
        addChannelsToGroup()

        pubnub.subscribe(
            channelGroups = listOf(expectedGroup),
            withPresence = true
        )

        retry {
            assertTrue(pubnub.getSubscribedChannelGroups().contains(expectedGroup))
        }
    }

    @Test
    fun testUnsubscribeFromChannelGroup() {
        addChannelsToGroup()

        pubnub.subscribe(
            channelGroups = listOf(expectedGroup),
            withPresence = true
        )

        wait()

        pubnub.unsubscribe(
            channelGroups = listOf(expectedGroup)
        )

        wait()

        assertTrue(pubnub.getSubscribedChannelGroups().isEmpty())
    }

    @Test
    fun testGetAllChannelsFrogroup() {
        addChannelsToGroup()

        pubnub.listChannelsForChannelGroup(
            channelGroup = expectedGroup
        ).await { result, status ->
            assertFalse(status.error)
            assertEquals(3, result!!.channels.size)
        }
    }

    @Test
    fun testAddChannelsToGroup() {
        pubnub.addChannelsToChannelGroup(
            channelGroup = expectedGroup,
            channels = listOf(expectedChannel1, expectedChannel2, expectedChannel3)
        ).await { _, status ->
            assertFalse(status.error)
            assertEquals(1, status.affectedChannelGroups.size)
            assertEquals(3, status.affectedChannels.size)
        }
    }

    private fun addChannelsToGroup() {
        pubnub.addChannelsToChannelGroup(
            channelGroup = expectedGroup,
            channels = listOf(expectedChannel1, expectedChannel2, expectedChannel3)
        ).await { _, status ->
            assertFalse(status.error)
        }
    }
}
